src/shims/unix/fs.rs

```diff
@@ -81,6 +81,89 @@ impl FileDescription for FileHandle {
+        &self,
+        communicate_allowed: bool,
+        op: FlockOp,
+    ) -> InterpResult<'tcx, io::Result<()>> {
+        assert!(communicate_allowed, "isolation should have prevented even opening a file");
+        #[cfg(target_family = "unix")]
+        {
+            use std::os::fd::AsRawFd;
+
+            // Transform `op` into host's flags
+            use FlockOp::*;
+            let host_op = match op {
+                SharedLock { nonblocking: false } => libc::LOCK_SH,
+                SharedLock { nonblocking: true } => libc::LOCK_SH | libc::LOCK_NB,
+                ExclusiveLock { nonblocking: false } => libc::LOCK_EX,
+                ExclusiveLock { nonblocking: true } => libc::LOCK_EX | libc::LOCK_NB,
+                Unlock => libc::LOCK_UN,
+            };
+
+            let fd = self.file.as_raw_fd();
+            let ret = unsafe { libc::flock(fd, host_op) };
+            let res = match ret {
+                0 => Ok(()),
+                -1 => Err(io::Error::last_os_error()),
+                ret => panic!("Unexpected return value from flock: {ret}"),
+            };
+            Ok(res)
+        }
+
+        #[cfg(target_family = "windows")]
+        {
+            use std::os::windows::io::AsRawHandle;
+            use windows_sys::Win32::{
+                Foundation::{ERROR_IO_PENDING, ERROR_LOCK_VIOLATION, FALSE, HANDLE, TRUE},
+                Storage::FileSystem::{
+                    LockFileEx, UnlockFile, LOCKFILE_EXCLUSIVE_LOCK, LOCKFILE_FAIL_IMMEDIATELY,
+                },
+            };
+            let handle = self.file.as_raw_handle() as HANDLE;
+
+            use FlockOp::*;
+            let mut lock_nb = false;
+            let ret = match op {
+                SharedLock { nonblocking } | ExclusiveLock { nonblocking } => {
+                    lock_nb = nonblocking;
+                    let mut flags = 0;
+                    if nonblocking {
+                        flags |= LOCKFILE_FAIL_IMMEDIATELY;
+                    }
+                    if let ExclusiveLock { .. } = op {
+                        flags |= LOCKFILE_EXCLUSIVE_LOCK;
+                    }
+                    unsafe { LockFileEx(handle, flags, 0, !0, !0, &mut std::mem::zeroed()) }
+                }
+                Unlock => unsafe { UnlockFile(handle, 0, 0, !0, !0) },
+            };
+
+            let res = match ret {
+                TRUE => Ok(()),
+                FALSE => {
+                    let mut err = io::Error::last_os_error();
+                    let code: u32 = err.raw_os_error().unwrap().try_into().unwrap();
+                    // Replace error with a custom WouldBlock error, which later will be mapped
+                    // in the `helpers` module
+                    if lock_nb && matches!(code, ERROR_IO_PENDING | ERROR_LOCK_VIOLATION) {
+                        let desc = format!("LockFileEx wouldblock error: {err}");
+                        err = io::Error::new(io::ErrorKind::WouldBlock, desc);
```

Comment:
This remapping is a bit hacky, but I couldn't find a better solution. Mapping it here or in `fd.rs` conflicts with mapping in `helpers.rs`.

Also, according to `LockFileEx` non-blocking call should return `ERROR_IO_PENDING`, but in practice we get `ERROR_LOCK_VIOLATION`, so to be safe I remap both of those.